File: tools/iibm_loader.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun 16 02:23:03 2015

@author: fer
"""
import numpy as np
from ConfigParser import ConfigParser


class iibm_loader():
    def __init__(self,reg_filename):
        self.reg_filename = reg_filename
        config = ConfigParser()
        config.read(reg_filename+'-0')
        
        self.fs = config.getfloat('GENERAL', 'fs')
        self.adc_scale = config.getfloat('GENERAL', 'adc_scale')
        self.channels = config.getint('GENERAL', 'channels')
        
        self.nfiles = config.getint('DATA_INFO','files')
        self.samples4file = config.getint('DATA_INFO','samples4file')
        self.samples4lastfile = config.getint('DATA_INFO','samples4lastfile')
        self.total_samples = self.samples4file * (self.nfiles - 1) + self.samples4lastfile
   
   
        self.dict_config = {}
        for section in config.sections():
            self.dict_config[section] = {}
            for key, val in config.items(section):
                self.dict_config[section][key] = val
        
        
        
    def get_data(self,channels = 'all', beg_time = 0, final_time='end'):
        if channels == 'all':
            channels = np.range(self.channels)
        
        beg_sample = int(np.floor(beg_time* self.fs))
        if self.total_samples < beg_sample:
            raise NameError('beg_time out of data')
            
        beg_file = int(beg_sample/self.samples4file)
        rel_beg_sample = beg_sample - beg_file * self.samples4file
        
        if final_time == 'end':
            final_sample = self.total_samples
        else:
            final_sample = int(np.ceil(final_time* self.fs))
        
        if final_sample <= beg_sample:
            raise NameError('final < beginning')
        
        final_file = int(final_sample/self.samples4file)
        rel_final_sample = final_sample - final_file * self.samples4file
        data = np.ndarray([len(channels),final_sample-beg_sample])
        if beg_file != final_file:
           #inicio:
            new_file = self.reg_filename + "-" + str(beg_file+1) #plus 1, because data begun in 1
            new_data = np.fromfile(new_file,np.int16)
            data[:, :self.samples4file - rel_beg_sample] = new_data.reshape([self.channels,new_data.size/self.channels],order='F')[channels,rel_beg_sample:]
            writed_samples = self.samples4file - rel_beg_sample
             
            #final
            new_file = self.reg_filename + "-" + str(final_file+1)#plus 1, because data begun in 1
            new_data = np.fromfile(new_file,np.int16)
            data[:, -rel_final_sample:] = new_data.reshape([self.channels,new_data.size/self.channels],order='F')[channels, :rel_final_sample]
        else:
            new_file = self.reg_filename + "-" + str(beg_file+1) #plus 1, because data begun in 1
            new_data = np.fromfile(new_file,np.int16)
            data = new_data.reshape([self.channels,new_data.size/self.channels],order='F')[channels,rel_beg_sample:rel_final_sample]
            writed_samples = self.samples4file - rel_beg_sample
       
 
        for i in range(beg_file+1,final_file+1): #50
            new_file = self.reg_filename + "-" + str(i+1)#plus 1, because data begun in 1
            new_data = np.fromfile(new_file, np.int16)
            data[:, writed_samples:writed_samples + self.samples4file]  = self.adc_scale * new_data.reshape([self.channels, new_data.size/self.channels], order='F')[channels, :]
            
        return data
if __name__ == '__main__':

    record = iibm_loader('test')
    if record.dict_config["GENERAL"]['format']=='example_raw':
        record.channels+=1

File: tools/iibm_loader.m
%This class require the fucntion ini2struct from:
%http://www.mathworks.com/matlabcentral/fileexchange/17177-ini2struct


classdef iibm_loader
    properties
        reg_filename
        fs
        adc_scale
        data_channels
        nfiles
        samples4file
        samples4lastfile
        total_samples
        struct_config
    end
    methods
        function obj = iibm_loader(FileName)
           obj.reg_filename = FileName;
           obj.struct_config = ini2struct(strcat(FileName,'-0'));
           obj.fs= str2double(obj.struct_config.general.fs);
           obj.adc_scale =  str2double(obj.struct_config.general.adc_scale);
           obj.data_channels =  str2double(obj.struct_config.general.channels);
           obj.nfiles =  str2double(obj.struct_config.data_info.files);
           obj.samples4file =  str2double(obj.struct_config.data_info.samples4file);
           obj.samples4lastfile =  str2double(obj.struct_config.data_info.samples4lastfile);
           obj.total_samples = obj.samples4file * (obj.nfiles - 1) + obj.samples4lastfile;
           
           
           obj.data_channels = obj.data_channels+1; %plus 1 non-electrode channel
        end
        
        function data = get_data(obj, channels, beg_time, final_time)
        %
		% Usage:
		%       channels is a vector from 1 to the data_channels,
		% with the selected channels

            switch nargin
                 case 1
                     channels = 'all';
                     beg_time = 0;
                     final_time= 'end';
                 case 2
                     beg_time = 0;
                     final_time= 'end';
                 case 3
                     final_time= 'end';
            end
            beg_sample = floor(beg_time* obj.fs);
            if obj.total_samples < beg_sample
				error('beg_time out of data.')
            end
            beg_file = floor(beg_sample/obj.samples4file);
            rel_beg_sample = beg_sample - beg_file * obj.samples4file +1;
            if strcmp(channels,'all')
               channels = [1:obj.data_channels];
            end
            if strcmp(final_time,'end')
               final_sample = obj.total_samples;
            else
                final_sample = ceil(final_time* obj.fs);
            end 
            if final_sample <= beg_sample
				error('final < beginning.')
            end
            final_file = floor(final_sample/obj.samples4file);
            rel_final_sample = final_sample - final_file * obj.samples4file +1;
            data = zeros(size(channels,2),final_sample-beg_sample);
            if beg_file ~= final_file
            %inicio:
                fid = fopen(strcat(obj.reg_filename,'-',num2str(beg_file+1)),'r'); %plus 1, because data began in 1
                data_aux = fread(fid,[obj.data_channels,obj.samples4file],'int16');
                data(:, 1:obj.samples4file - rel_beg_sample) = data_aux(channels,rel_beg_sample:end);
                writed_samples = obj.samples4file - rel_beg_sample+1; %begins in the first element dont writed
                fclose(fid);

                %final
                fid = fopen(strcat(obj.reg_filename, '-', num2str(final_file + 1)), 'r'); %plus 1, because data began in 1
                data_aux = fread(fid, [obj.data_channels,obj.samples4file], 'int16');
                data(:, end - rel_final_sample:end) = data_aux(channels,1:rel_final_sample);
                fclose(fid);
		
                for i = [beg_file+1,final_file+1]
                    fid = fopen(strcat(obj.reg_filename, '-', num2str(i + 1)), 'r'); %plus 1, because data began in 1
                    data_aux = fread(fid,[obj.data_channels,obj.samples4file],'int16');
                    data(:, writed_samples:writed_samples+obj.samples4file) = data_aux(channels,:);
                    

                    writed_samples = writed_samples+obj.samples4file+1; %begins in the first element dont writed
                    fclose(fid);
                end
				
            else

                fid = fopen(strcat(obj.reg_filename, '-', num2str(beg_file + 1)), 'r'); %plus 1, because data began in 1
                data_aux = fread(fid,[obj.data_channels,obj.samples4file],'int16');
                data = data_aux(channels,rel_beg_sample:rel_final_sample);          
            
            end

        end
        
    end
end


